File: contests/lc/3606.CouponCodeValidator/solution.go
package main

import (
	"regexp"
	"sort"
)

// validateCoupons filters valid coupon codes and returns them sorted by business line
// (electronics, grocery, pharmacy, restaurant) and then lexicographically by code.
func validateCoupons(code []string, businessLine []string, isActive []bool) []string {
	order := map[string]int{
		"electronics": 0,
		"grocery":     1,
		"pharmacy":    2,
		"restaurant":  3,
	}
	validCode := regexp.MustCompile(`^[A-Za-z0-9_]+$`)

	type coupon struct {
		code string
		biz  string
	}

	var coupons []coupon
	for i := range code {
		biz := businessLine[i]
		if !isActive[i] {
			continue
		}
		if _, ok := order[biz]; !ok {
			continue
		}
		if code[i] == "" || !validCode.MatchString(code[i]) {
			continue
		}
		coupons = append(coupons, coupon{code: code[i], biz: biz})
	}

	sort.Slice(coupons, func(i, j int) bool {
		if order[coupons[i].biz] == order[coupons[j].biz] {
			return coupons[i].code < coupons[j].code
		}
		return order[coupons[i].biz] < order[coupons[j].biz]
	})

	result := make([]string, len(coupons))
	for i, c := range coupons {
		result[i] = c.code
	}
	return result
}

File: contests/lc/3606.CouponCodeValidator/solution.py
import re
from typing import List


class Solution:
    def validCouponCodes(
        self, code: List[str], businessLine: List[str], isActive: List[bool]
    ) -> List[str]:
        order = {
            "electronics": 0,
            "grocery": 1,
            "pharmacy": 2,
            "restaurant": 3,
        }
        valid_code = re.compile(r"^[A-Za-z0-9_]+$")

        coupons = []
        for c, b, active in zip(code, businessLine, isActive):
            if not active:
                continue
            if b not in order:
                continue
            if not c or not valid_code.fullmatch(c):
                continue
            coupons.append((order[b], c))

        coupons.sort(key=lambda x: (x[0], x[1]))
        return [c for _, c in coupons]
